### DISCUSSION TITLE
Stop names to touch display.

### DISCUSSION THREAD

**Discussion by Dinsky9898** [GrandOrgue/grandorgue#1675] [Developers]
Hi, folks!
I planning my new organ and will use touch displays on the jambs. Are there any way to extract the stop info from GO when starting or when changing an organ sample ? I am using Arduino Due via USB from my Windows 11 PC, via multiplexers and to/from touch display.
Thanks....

**Comment by larspalo** [comment]
Touch displays that work under the current OS used will also work for GO usage. Many recent sample sets have custom panels for the stop jambs (usually one for the left and another one for the right) that are intended for just such touch screen setups. Creating additional jamb panels, for sample sets that doesn't have them by default, can be done with for instance with https://github.com/larspalo/GOODF with some basic work and a little knowledge.

**Reply by Dinsky9898** [reply]
Thanks, Lars.
The idea came from Kenneth Spensers videos: https://www.youtube.com/watch?v=inbJSIQWzgQ&list=PL3lNsK1BBlTk4JZkV61VWOTC5evPm5pYo
Insted of buttons I would like to use touch displays like for instance : https://www.aliexpress.com/item/1005004241997895.html?spm=a2g0o.order_list.order_list_main.342.49391802y3ZuLQ

**Comment by midimusic** [comment]
There is no need to extract any information from GO. You can read the stop names from the computer screen - as far as I am aware the ODF's contain stop numbers rather than names. Easy to check!

As for setting up - you can feed GO any sensible midi control codes and set up GO to use these codes for setting stops. You can use the same codes on all samplesets. Each sample set must be individually set up to use the codes.

This is obviously a lot of work but you appear to like this kind of work so good luck. (Lars method would work just as well in my opinion and involve minimal work from you).

csw900

**Reply by davidgritter** [reply]
I assume you are using arduino(s) to drive these individual displays and you want to display the names of the stops from the current odf on individual displays for each drawknob.  In that  case you can edit each drawstop in the odf and tell it to send a 16 bit or 32 bite sysex value message (see figure.)  Each drawknob has to send a unique ID  (and may also send a color.
If you are using the standard arduino midi library, you can program midi to automatically call a handler when a sysex message is received with, for example:
Organ.setHandleSystemExclusive(OnSysex); 
where Organ is an instance of a MIDI interface

The sysex messages can be decoded by the arduino to recieve the stop name, for example:

void OnSysex(byte* message, unsigned len) {
  // entire message is included including leading f0 and trailing 57
  int messID, messColor;
  char label[32];
  char nextxt[32];
  //Serial.println("sysex received");
  if (message[1] == 0x7D) {
    if (message[2] == 0x19) {  // 16 bit hauptwerk message
      messID = message[3];
      parsetext(message, 4, len, label);
      messColor = 0;
      if (messID == 102) {  //memory level
       memlev.setText(label,true);
      } else if (messID == 103) {  //transpose semitones
        transpose.setText(label,true);
      }
    } else if (message[2] == 0x01) {  // 32 bit hauptwerk message
      messID = 128 * message[3] + message[4];
     // Serial.println(messID);
      messColor = message[5];
      parsetext(message, 6, len, label);
      if (messID == 100) {  // message from dynamic organ label on Master panel
        splittext(label,nextxt,8,20);
        organName.setText(nextxt,true);
        
      }

      else if (messID == 101) {  //message from dynamic label on temperament panel
       splittext(label,nextxt,12,24);
       temperLabel.setText(nextxt,true);
       //Serial.println("temperament message sent");
      }
    }
  }
}
the following parses out the text (call to parsetext above it assumes that two ascii spaces in a row mean the string is complete:

void parsetext(byte* mess, int start, int len, char* label) {
  for (int n = start; n<len; n++) {
    if (mess[n] == 0xF7) {
      label[n - start] = 0;
      break;
    }
    if (mess[n] == 0x20) {
      if ((mess[n + 1] == 0x20) || (mess[n + 1] == 0xF7)) {
        label[n - start] = 0;
        break;
      } 
      
    }
    label[n - start] = mess[n];
    
  }
}
calls to: 
splittext(label,nextxt,8,20);
organName.setText(nextxt,true);
are unique to my display and should be replaced with whatever code writes to your displays.

![Screenshot_20231016_140509](https://github.com/GrandOrgue/grandorgue/assets/94871928/e56ca3e7-53c7-4067-8d6f-7d71b719bd45)

**Reply by Dinsky9898** [reply]
Very interesting, Davidgritter !! 
Thank you. This is just what I'm need I think. I will try it out.

**Reply by davidgritter** [reply]
Note, that stops behave differently than labels when sending Sysex.  Therefore for a stop drawknob object the message needs to be
a name message, not the value message shown in the graphic above.  For stops the value sends on or off, and name sends the stop name

**Comment by willeke1234** [comment]
Interesting subject. As far as I know, GO can't send stop names. GO can send the name of the organ and can load an organ when receiving a MIDI event. It is possible to synchronize GO and a Stop Label app.

Related:
Can GO output Stop Name text via MIDI SysExRaw messages? #407
[RFC] Add an API to call GrandOrgue #1563

@davidgritter You use active labels in your code. How do you get buttons to send their names?

**Reply by midimusic** [reply]
Surely the only place that stop name text ever appears in GO is on the stops themselves. Thus the GO code has no knowledge of these names and can never send them.
csw900

**Reply by willeke1234** [reply]
@midimusic The GO code knows the names. The names are displayed in the MIDI Objects list and the MIDI event dialog title.

**Reply by Dinsky9898** [reply]
Shure. So we have here what we need. Sending it via SysEx should be a simple task. A switch in the Organ Settings could tell if these messages should be sent or not.

**Reply by willeke1234** [reply]
@Dinsky9898 How do you know which name goes to which display?

**Reply by davidgritter** [reply]
It is not true that GrandOrgue doesn't know the stop names.   When right clicking on a drawknob, you are setting up the send and receive midi messages for that GUI element, not for the stop or rank which is separately defined in the ODF.  For example, the stopknob could be separated from and actual ODF stop by a whole series of Switch elements  @willeke1234 : See the sysex message in my initial post above.  For Stopknobs, value sends on/off, while name sends the text that appears on the stopknob, or that is used when creating the GUIelement in the ODF if a graphical stopknob is created. (see lars Palo's GOODF for  more info on how GUIelements are created)  For some ODFs the GUIelement names are not particularly meaningfull or have extra information, but these can always be changed with GOODF.  I have been using Stop name sysex messages for a long time to make user interfaces with nextion displays and Arduinos with USBmidi capability. The Arduino code associates the sysex ID (see popup  with a particular hardware display element @Dinsky9898 There is no need for any additional switch in the organ settings.  The Sysex messages are always sent unconditionally  at an appropriate time (loading or reseting an organ, or when a value changes) if you have defined them.  Nothing is sent if you haven't defined a sysex for a particular graphical element. So setting up the sysex messages is exactly the same as the way we set up a GUI element to respond to a MIDI noteon/noteoff from a hardware stop switch, or to send a noteon/noteoff to turn on a hardware LED  or trigger a coil in an electrically activated drawknob or tablet.

**Reply by willeke1234** [reply]
@davidgritter  Aha, I always wondered what those MIDI events are supposed to do, they aren't in the help and aren't sent when the value changes. After reading this: Can GO output Stop Name text via MIDI SysExRaw messages? #407, I thought sending stop names isn't possible.

It looks like GO sends the `Name` field (name in MIDI objects) and not the optional `DispLabelText` field (GUI).

**Reply by davidgritter** [reply]
Exactly, The  Sysex messages send the Name of the GUIelement as seen in the GOODF screenshot below, along the left edge, not the DisplayLabelText entered in the main window.   If you don't like  the GUI element name, it is inhereted from the name of the stop or switch element which they control.  These names can arbitrarily be changed in GOODF,  by editing the associated switch or stop, and it is no problem for multiple elements to have the same name.  As soon as you change the name of a switch, for example, in GOODF, the name of its associated GUI element immediately changes as well.
As far as I know, there are no sysexraw messages that can be programmed in Grandorgue, other than the four Hauptwerk sysex messages that already exist.
![image](https://github.com/GrandOrgue/grandorgue/assets/94871928/4c0c012d-3ff3-4d85-b011-79b71ce7357a)
